src/instances.ts

```diff
@@ -302,6 +302,47 @@ function getExistingSolutionBuilderHost(key: FilePathKey) {
   return undefined;
 }
 
+const addAssetHooks = (
+  loader: webpack.loader.LoaderContext,
+  instance: TSInstance
+) => {
+  // Adding assets in afterCompile is deprecated in webpack 5 so we
+  // need different behavior for webpack4 and 5
+
+  if (!!webpack.version!.match(/^4.*/)) {
```

Comment:
Do you think we should do caching here for perf reasons? The webpack version value won't change so regex-ing each time is perhaps less performant than if we cached it?

Reply:
This is only run once per compile cycle so I'm not sure how much difference it makes.  I understand javascript regexes are not fast but I'm not sure if the impact is significant here.  We could eliminate the regex by using
```
if (!loader._compilation.hooks.afterProcessAssets)
```
instead, as this does not exist in webpack4, but I thought checking webpack.version makes the code easier to understand.  I assume that in the medium term there will be a webpack5 version of ts-loader so this is probably not a long term issue.

If you think caching is worthwhile I'll update the PR.

Reply:
as an alternative to testing each time how about doing something like this:

```
// Adding assets in afterCompile is deprecated in webpack 5 so we
// need different behavior for webpack4 and 5
const addAssetHooks = !!webpack.version!.match(/^4.*/) ? (
  loader: webpack.loader.LoaderContext,
  instance: TSInstance
) => {
    // add makeAfterCompile with addAssets = true to emit assets and report errors
    loader._compiler.hooks.afterCompile.tapAsync(
      'ts-loader',
      makeAfterCompile(instance, true, true, instance.configFilePath)
    );
}
: (
  loader: webpack.loader.LoaderContext,
  instance: TSInstance
) => {
  // We must be running under webpack 5+

  // Add makeAfterCompile with addAssets = false to suppress emitting assets
  // during the afterCompile stage. Errors will be still be reported to webpack
  loader._compiler.hooks.afterCompile.tapAsync(
    'ts-loader',
    makeAfterCompile(instance, false, true, instance.configFilePath)
  );

  // Emit the assets at the afterProcessAssets stage
  loader._compilation.hooks.afterProcessAssets.tap('ts-loader', (_: any) => {
    makeAfterCompile(
      instance,
      true,
      false,
      instance.configFilePath
    )(loader._compilation, () => {
      return null;
    });
  });

  // It may be better to add assets at the processAssets stage (https://webpack.js.org/api/compilation-hooks/#processassets)
  // This requires Compilation.PROCESS_ASSETS_STAGE_ADDITIONAL, which does not exist in webpack4
  // Consider changing this when ts-loader is built using webpack5
};
```

Reply:
That is a neat solution. Are you going to make the change or do you want me to update the PR?

Reply:
I'm not on my computer right now - so if you're able to do it that'd be amazing!  But yeah - I'd be happy to do at some time over the weekend if you're busy.  Happy either way